internal/topology/node.go

```diff
@@ -0,0 +1,222 @@
+package topology
+
+import (
+	"context"
+	"net/url"
+	"sync/atomic"
+	"time"
+
+	"github.com/contentsquare/chproxy/internal/counter"
+	"github.com/contentsquare/chproxy/internal/heartbeat"
+	"github.com/contentsquare/chproxy/log"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	// prevents excess goroutine creating while penalizing overloaded host
+	DefaultPenaltySize     = 5
+	DefaultMaxSize         = 300
+	DefaultPenaltyDuration = time.Second * 10
+)
+
+type nodeOpts struct {
+	defaultActive   bool
+	penaltySize     uint32
+	penaltyMaxSize  uint32
+	penaltyDuration time.Duration
+}
+
+func defaultNodeOpts() nodeOpts {
+	return nodeOpts{
+		penaltySize:     DefaultPenaltySize,
+		penaltyMaxSize:  DefaultMaxSize,
+		penaltyDuration: DefaultPenaltyDuration,
+	}
+}
+
+type NodeOption interface {
+	apply(*nodeOpts)
+}
+
+type penalty struct {
+	penaltySize     uint32
+	penaltyMaxSize  uint32
+	penaltyDuration time.Duration
+}
+
+func (o penalty) apply(opts *nodeOpts) {
+	opts.penaltySize = o.penaltySize
+	opts.penaltyMaxSize = o.penaltyMaxSize
+	opts.penaltyDuration = o.penaltyDuration
+}
+
+func WithPenalty(size, max uint32, duration time.Duration) NodeOption {
+	return penalty{
+		penaltySize:     size,
+		penaltyMaxSize:  max,
+		penaltyDuration: duration,
+	}
+}
+
+type defaultActive struct {
+	active bool
+}
+
+func (o defaultActive) apply(opts *nodeOpts) {
+	opts.defaultActive = o.active
+}
+
+func WithDefaultActiveState(active bool) NodeOption {
+	return defaultActive{
+		active: active,
+	}
+}
+
+type Node struct {
+	// Node Address.
+	addr *url.URL
+
+	// Whether this node is alive.
+	active atomic.Bool
+
+	// Counter of currently running connections.
+	connections counter.Counter
+
+	// Counter of unsuccesfull request to decrease host priority.
+	penalty atomic.Uint32
+
+	// Heartbeat function
+	hb heartbeat.HeartBeat
+
+	// For metrics only
+	clusterName string
```

Comment:
Cluster name and replica name mostly seem to be used for logging & metrics. However, replica name is also used outside of the package. Maybe at some point we can remove this, the `Node` doesn't have to be aware of the cluster & replica it is in for chproxies purposes